### PR TITLE
RFC 0007 Support tabs for indentation.

### PR DESCRIPTION
**Pull request by ioquatix**
For context see <https://github.com/yaml/yaml-spec/issues/40>.

### CODE REVIEW

**Comment by ingydotnet** [comment]
Samuel (@ioquatix),

Thanks for your interest in YAML and for taking the time to make a PR.

Unfortunately at this time the [RFC process is on hold](https://github.com/yaml/yaml-spec/wiki/YAML-RFC-Process) while we finish up the [1.2.1 spec publication](https://github.com/yaml/yaml-spec/wiki/YAML-1.2.1-Specification).

We have a large backlog of RFCs including one for tab indentation support. This should all become apparent in the not to distant future.

I'm closing this for now but will take it into consideration at the appropriate time. I'll make sure you are notified when that time arrives.

**Comment by ioquatix** [comment]
Okay, sounds good.

**Comment by ioquatix** [comment]
I see that 1.2.1 was published. Well done. What's the next step for the RFC process?

**Comment by ingydotnet** [comment]
I think you mean 1.2.2. :)
We renamed it after you were last here.
We also renamed RFC to YEP.
We are actively working on that right now, and getting the kinks out before it goes public.
Hang in there!

**Comment by ioquatix** [comment]
Thanks for the update!

**Comment by ingydotnet** [comment]
@kachkaev At this point work on the YAML specification is on hold again.
I suppose I would advise you to not have any expectations of a YAML spec with normative changes being available this year.
(Note: I'm currently working on stuff that may help you here, but it's too soon to even discuss.)

The point on accessibility is a fair one, but I'm not convinced that tabs can be added to YAML.
One specific issue is that the numeric explicit indentation indicator for folded and literal block scalars is visually columnar with spaces, and with tabs, What You See Is Not What You Get. Granted it's a fairly rare part of the syntax, but there may be other issues too.

To be clear, indentation with mixed spaces and tabs can never be possible. If tabs are permitted for indentation, they must be used for the entire stream (or at least the entire document). Otherwise the semantics could easily be ambiguous.

Come to think of it, re [the post you referenced](https://www.reddit.com/r/javascript/comments/c8drjo/nobody_talks_about_the_real_reason_to_use_tabs/) talks about a tabwidth of 1 column, but that won't work in YAML for at least this reason: https://play.yaml.io/main/parser?input=LQogIC0KICAgIC0gZm9vCi0gLSAtIGJhcg==

You can't do `- - - bar` with tabs and use a tabwidth of 1.

Circa 2004 the YAML core devs spent 3 months batting around every idea to try to include tabs. It's all in the yaml-core mailing list archives if you feel like digging into the past. By the end we were pretty confident that YAML without tabs was the best choice.

To recap, a new spec ain't coming in any reasonable to work with timeframe. It's far from certain that tabs will make it into the language.
But it's also far from certain that they won't :)

**Comment by ioquatix** [comment]
> You can't do - - - bar with tabs and use a tabwidth of 1.

Just limit it to indentation after a newline?

**Comment by ingydotnet** [comment]
I think you misunderstood me.
You can do this in YAML:
```
-   -   - foo
        - bar
    - baz
```
the spaces between the dashes are indentation.

In YAML with tab indentation, the spaces would be tabs (set to tabwidth 4.
With tabwidth 1, you get:
```
- - - foo
  - bar
 - baz
 ```
 which obviously doesn't work.
 
 Tabwidth 2 (or more) is fine
 ```
 - - - foo
     - bar
   - baz
```

**Comment by ingydotnet** [comment]
FWIW here's a simple example where tabwidth makes yaml ambiguous when you allow mixed tabs and space for indentation:
```
a:
  b:
    c:
      d:
        e:
--->f: g
```
Given that `--->` this YAML could be valid and mean different things for tabwidth 2, 4, 6, 8, and >8.

So indentation needs to be all spaces or all tabs.

With all tabs you have this issue...
Assume the following YAML uses tabs for indent but you can't see them:
```
foo: |2
          bar
        baz
```
Unless you know what the tabwidth for this visual rendering is here, the value is ambiguous.

BTW Tabs are allowed all over the place in YAML 1.2. Even at the start of lines. Just not when it counts as indentation.
See this example: https://play.yaml.io/main/parser?input=CSMgY29tbWVudAoJCQkKZm9vOgliYXIKIAliYXIKYmF6OgotCTEKLQkJMgo=
This is valid YAML (although half the parsers tested can't parse it) and the tabs are visible.

**Comment by tanriol** [comment]
Python 3, which also has significant indentation levels, has a rule that indentation has to be consistent, meaning that if the indentation level of a line depends on the tab width, the file is invalid. I think that basically boils down to tracking indentation as "N tabs + M spaces", which leaves the obvious way of comparing two indents to say whether the second is greater, less, equal or inconsistent with the first.

**Comment by ioquatix** [comment]
> Given that ---> indicates a single tab, this YAML could be valid and mean different things for tabwidth 2, 4, 6, 8, and >8.

Why? Surely a single tab = a single unit of indentation. In the above example a single level of indentation is equivalent to 2 spaces. There is no situation where a tab should be interpreted as more than one unit of indentation. I don't think anyone is expecting the behaviour to change depending on the tab width. I assume the parser is keeping track of indentation levels, and a single tab = a single indentation level increment. Where as I assume for spaces it depends on the amount of preceeding whitespace, e.g. when you detect 2 or more spaces, you consume all spaces and consider that to be one level of indentation. When you encounter 1 tab, you consider that one unit of indentation.

> So indentation needs to be all spaces or all tabs.

Yes, you could do that, and you could even define a tab to be equivalent to some fixed level of indentation too as a way of avoiding dealing with this in the parser. But it's not that HARD IMHO given you already handle variable number of spaces for indentation. As mentioned by @tanriol Python handles this pretty well. But I've also done this in the past when parsing nested data structures. It boils down to an algorithm like this:

```ruby
def parse
  current_indentation = ""
  level = 0
  
  while parse_line in {indentation:, statement:}
    if indentation.start_with?(current_indentation)
      if indentation.size > current_indentation.size
        # The indentation level was increased:
        current_indentation = indentation
        level += 1
      else
        # The level is the same.
      end
    elsif current_indentation.start_with?(indentation)
      if indentation.size < current_indentation.size
        # The indentation level was decreased:
        current_indentation = indentation
        level -= 1
      else
        # The level was the same.
      end
    else
      raise ParserError, "Inconsistent indentation detected!"
    end
    
    parse_statement(level, level)
  end
end
```

If you treat a tab as a single unit of indentation I don't think you can really go wrong. Trying to treat a tab as some arbitrary number of spaces is never going to work and you shouldn't go down that path IMHO.

**Comment by tanriol** [comment]
Not that easy - you need to maintain indentation stack as a single line can drop more than one level of indentation.

**Comment by ioquatix** [comment]
@tanriol yep you are quite right, and I assume YAML already does this since it allows an arbitrary number of spaces for indentation levels, e.g.

```yaml
foo:
          bar1:
            baz: 10
          bar2:
              baz: 10
```

parses totally fine for me.

**Comment by ioquatix** [comment]
@ingydotnet any way we can make progress here?

**Comment by ingydotnet** [comment]
@ioquatix Work on a 1.3 spec is on hold, likely at least until 2023.

Personally I am not currently a proponent of tab indentation for YAML, but if/when we institute a public RFC process for the spec, it would certainly be discussed.

On a practical note, maybe you want to work on a YAML pre-processor that converts tab indentation to space indentation.
This could be good for seeing if it would really work out well across the board.

**Comment by ioquatix** [comment]
https://github.com/prettier/prettier/issues/7475

**Comment by ioquatix** [comment]
> The main reason I would like to see this change is for refreshable braille displays that are used by blind programmers a lot. Each space wastes one braille cell and takes away valuable braille realestate. So if the default indentation of a project is 4 spaces per level, a 3rd level indentation wastes 12 braaille cells before code starts. On a 40 cell display, which is the most commonly used with notebooks, this is more than a quarter of the available cells wasted with no information. If each indentation level was represented by only one tab character, there would be three cells occupied by a tab character each, and on the 4th cell, the code would start. That's less than 10 percent occupied on the same length display, but all cells contain valuable information that is easily discoverable and immediately comprehensible.

— [MarcoZehe on GitHub](https://github.com/prettier/prettier/issues/7475#issuecomment-668544890) in a PR requesting that prettier change their useTabs setting from default false to default true.

**Comment by ioquatix** [comment]
@ingydotnet any update/news on the RFC process/YAML 1.3 spec discussions?

**Comment by ingydotnet** [comment]
Thanks for the reminder. I'll have some exciting news to share soon. Remind me again if you don't hear back by April.

**Comment by kripper** [comment]
> Thanks for the reminder. I'll have some exciting news to share soon. Remind me again if you don't hear back by April.

You are reminded.

**Comment by kripper** [comment]
I forked PyYAML and added tabs indentation support here:
https://github.com/kripper/pyyaml

**Comment by ioquatix** [comment]
@ingydotnet any update on this proposal, YEP, etc?
